### PR TITLE
BLOCKCHAIN-228 - FE - for action buttons like "Blacklist" or "Repeal referendum" - detect if there's already an ongoing motion/referendum

### PR DESCRIPTION
**Pull request by MichalLustykNeti**
image

### CODE REVIEW

File: src/api/nodeRpcCall.js [hunk 1]
```diff
@@ -1844,6 +1861,8 @@ const congressDemocracyBlacklist = async (proposalHash, referendumIndex, walletA
 
   const threshold = await congressMajorityThreshold();
   const proposal = api.tx.democracy.blacklist(proposalHash, referendumIndex ?? null);
+
+  console.log({proposalHash, referendumIndex })
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:1865]:
```suggestion
```

File: src/api/nodeRpcCall.js [hunk 2]
```diff
@@ -1905,6 +1924,21 @@ const congressAmendLegislationViaReferendum = async (
   return await congressProposeReferendum(amendLegislation, fastTrack, votingPeriod, enactmentPeriod, walletAddress);
 }
 
+const loadProposalsBlacklistMotions = async (proposals, isReferendum) => {
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:1927]: Is this unused...?

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]
This issue is about all actions that create motions/referenda, not just blacklists. At the very least, its missing support for repealing legislation motions/referenda.

**Comment by kacperzuk-neti** [comment]
Fix commit msg.

File: src/components/Voting/Referendum/Items/ProposalItem/index.js [hunk 3]
```diff
@@ -148,8 +159,13 @@ ProposalItem.propTypes = {
   description: PropTypes.string.isRequired,
   userDidEndorse: PropTypes.bool.isRequired,
   boundedCall: call.isRequired,
-  buttonEndorseCallback: PropTypes.bool.isRequired,
-  proposalIndex: PropTypes.number.isRequired,
+  buttonEndorseCallback: PropTypes.func.isRequired,
+  proposalIndex: PropTypes.string.isRequired,
+  blacklistMotion: PropTypes.string,
```
Comment (kacperzuk-neti) [inline src/components/Voting/Referendum/Items/ProposalItem/index.js:164]: It should be required and contain `null` if there's no motion.

File: src/components/Voting/Referendum/Items/ReferendumItem/index.js [hunk 4]
```diff
@@ -203,6 +217,11 @@ ReferendumItem.propTypes = {
   votingTimeLeft: PropTypes.string.isRequired,
   referendumIndex: PropTypes.number.isRequired,
   delegating: PropTypes.bool.isRequired,
+  blacklistMotion: PropTypes.string,
```
Comment (kacperzuk-neti) [inline src/components/Voting/Referendum/Items/ReferendumItem/index.js:220]: It should be required and contain null if there's no motion.

File: src/redux/actions/congress.js [hunk 5]
```diff
@@ -22,6 +22,7 @@ export const {
   renounceCandidacy,
   unapproveTreasurySpend,
   voteAtMotions,
+  citizenProposeRepealLegislation,
```
Comment (kacperzuk-neti) [inline src/redux/actions/congress.js:25]: This has nothing to do with congress. It should be either `democracy` or `legislation` actions (put it wherever we already have the action for proposing new legislation)

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

File: src/api/nodeRpcCall.js [hunk 6]
```diff
@@ -925,36 +942,117 @@ const revertVetoForLegislation = async (tier, id, section, walletAddress) => {
...
+  const motions = (await api.query.council.proposals()).map((propose) =>
+    propose.toString()
+  );
+  const publicProps = (await api.query.democracy.publicProps())
+    .toJSON()
+    .map((proposal) => proposal[1].lookup.hash ?? proposal[1].legacy?.hash);
+  
+  const referendums = await api.query.democracy.referendumInfoOf.entries();
+  const referendumProposals = referendums
+    .map(([_, exposure]) => {
```
[the remark this answers is not in the document]
Comment (MichalLustykNeti) [reply src/api/nodeRpcCall.js:1003]: from polka docs it may be reanemed

[the remark this answers is not in the document]
Comment (kacperzuk-neti) [reply src/api/nodeRpcCall.js:1003]: Docs for what? Can you link them?

File: src/components/Modals/CitizenRepealLegislationModal.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    >
+      <div className={styles.h3}>Propose a Congress Motion - propose referendum for legislation repeal</div>
+
+      <div className={styles.title}>Legislation Tier</div>
+      <SelectInput
+        register={register}
+        name="tier"
+        disabled
+        options={[
+          { value: 'InternationalTreaty', display: 'International Treaty' },
```
[the remark this answers is not in the document]
Comment (MichalLustykNeti) [reply src/components/Modals/CitizenRepealLegislationModal.js:50]: I coppy that from CongressRepealLegislationFastTrackModal i fot that  we would like to use here only these laws. (i now added here Constitution also)

File: src/api/nodeRpcCall.js [hunk 8]
```diff
@@ -925,36 +942,117 @@ const revertVetoForLegislation = async (tier, id, section, walletAddress) => {
...
+          if (referendumProposals.includes(mainRepealLawHash))
+            legislationById[year][index].repealReferendum = mainRepealLawHash;
+          if (publicProps.includes(mainRepealLawHash))
+            legislationById[year][index].repealProposal = mainRepealLawHash;
+          await Promise.all(
+            sections
+              .map((v) =>
+                v.content.isSome ? v.content.unwrap().toHuman() : "Repealed"
+              )
+              .map(async (_, section) => {
```
[the remark this answers is not in the document]
Comment (MichalLustykNeti) [reply src/api/nodeRpcCall.js:1031]: Yes to get repeal law hash we do not need law content so se simply drop the content of it

File: src/api/nodeRpcCall.js [hunk 9]
```diff
@@ -1779,7 +1877,7 @@ const congressProposeReferendum = async (
   const threshold = await congressMajorityThreshold();
 
   const proposeAndVote = await createProposalAndVote(threshold, proposal, true)
-  const existingPreimage = await api.query.preimage.preimageFor([proposal.hash, proposal.encodedLength])
+  const existingPreimage = await api.query.preimage.preimageFor([referendumProposal.hash, referendumProposal.encodedLength])
```
[the remark this answers is not in the document]
Comment (MichalLustykNeti) [reply src/api/nodeRpcCall.js:1883]: referendumProposal is passed to func as argument
image

File: src/components/Congress/CitizenProposeRepealLegislationButton/index.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import CitizenRepealLegislationModal from '../../Modals/CitizenRepealLegislationModal';
+import {
+  congressSelectors,
+} from '../../../redux/selectors';
+
+export default function CitizenProposeRepealLegislationButton({ tier, id, section }) {
+  const [isModalOpen, setIsModalOpen] = useState(false);
+  const handleModalOpen = () => setIsModalOpen(!isModalOpen);
+
+  // requires parent to dispatch getMembers action
```
[the remark this answers is not in the document]
Comment (MichalLustykNeti) [reply src/components/Congress/CitizenProposeRepealLegislationButton/index.js:14]: Yes it is it uses the same logic as RepealLegislationButton
image

File: src/components/Legislation/LegislationView/index.js [hunk 11]
```diff
@@ -97,6 +135,7 @@ function ActionButtons({ tier, id, section }) {
         <CongressAmendLegislationViaReferendumButton {...{ tier, id, section }} />
       </>
       )}
+      { (repealReferendum || repealProposal) && <CitizenProposeRepealLegislationButton {...{ tier, id, section }} />}
```
[the remark this answers is not in the document]
Comment (MichalLustykNeti) [reply src/components/Legislation/LegislationView/index.js]: True i changed it already

**Pull request closed without merging**